Mps2 an521 and Musca B1 support call TFM services by IPC

karl-zh opened the pull request:
sample: tfm_integration: Zephyr plus TF-M IPC mode
    
TF-M supports call secure services by IPC (Inter-Processes
Communication) mode. All services expose to non-secure by the PSA APIs.
IPC mode is aims to comply the requirment of PSA-FF. It supports
asynchronous access from non-secure to secure.
This patch allows Zephyr in non-secure to call TF-M services by PSA
APIs on AN521. When this changes get approved, will have Musca B1 supported later on, because B1 needs migrate to eFlash to align with TFM, now the B1 related patches tracking by https://github.com/zephyrproject-rtos/zephyr/pull/16729.

BL2 is enabled due to the PSA isolation level 1 requirment, tfm_s.bin
and zephyr.bin should be signed.

The test cases are originally from TF-M (ipc_ns_interface_testsuite.c).

Commit:
sample: tfm_integration: Zephyr plus TF-M IPC mode

TF-M supports call secure services by IPC (Inter-Processes
Communication) mode. All services expose to non-secure by the PSA APIs.
IPC mode is aims to comply the requirment of PSA-FF. It supports
asynchronous access from non-secure to secure.
This patch allows Zephyr in non-secure to call TF-M services by PSA
APIs on AN521.

BL2 is enabled due to the PSA isolation level 1 requirment.

The test cases are originally from TF-M (ipc_ns_interface_testsuite.c).

TF-M Commit: 5358c8c4cf385cde6091adc74dce53fa3f0c2ce4

Signed-off-by: Karl Zhang <karl.zhang@linaro.org>

dbkinder commented on samples/tfm_integration/README.rst:
```diff
@@ -0,0 +1,61 @@
+
+Overview
+********
+A simple TF-M integration example that can be used with arm v8-M supported
```
complete sentence please:
```
This is a simple TF-M integration example that can be used with an Arm v8-M
supported board.
```

dbkinder commented on samples/tfm_integration/README.rst:
```diff
@@ -0,0 +1,61 @@
+********
+A simple TF-M integration example that can be used with arm v8-M supported
+board.
+To run sample for TF-M in secure and Zephyr in non-secure, Zephyr requsets
```
How about:
```
Zephyr uses TF-M Platform Security Architecture (PSA) APIs to run
this sample in a secure configuration with
Zephyr itself in a non-secure configuration. The sample prints test info to the
console either as a single- or multi-threaded application.
```
and delete the two bullet points.

dbkinder commented on samples/tfm_integration/README.rst:
```diff
@@ -0,0 +1,61 @@
+This project outputs test status and info to the console. It can be built and executed
+on MPS2+ AN521 as follows:
+
+1. Build Trusted Firmware (TF-M) without Bootloader.
```
I'm not sure these instructions are sufficient for a developer to follow (do we need to provide the steps, for example, for how to build TF-M without a bootloader or build Zephyr with a non-secure configuration)?

I think the detailed instructions are in the board document, right? If so, let's update the intro sentence to link to that document:

```
This sample project outputs test status and info to the console.
It can be built and executed on MPS2+ AN521 as follows (see the
:ref:`mps2_an521_board` board document for instruction details):
```

dbkinder commented:
You'll also need to add tfm_integration to the list of samples in /samples/index.rst (to fix the toctree error in the document build)

dbkinder reviewed (approved):
OK for docs.

karl-zh commented on samples/tfm_integration/README.rst:
```diff
@@ -0,0 +1,88 @@
+********************
+
+This project outputs test status and info to the console. It can be built and
+executed on MPS2+ AN521, Musca B1.
```
TODO: Better to add an item to list specific command lines about how to build TFM per the suggestion from @microbuilder, to make it clearer for anyone to run this sample easily.

microbuilder reviewed (changes requested):
I've tested this and it seems good to me other than some documentation issues.

Needs to be rebased against master, and I will test again with QEMU 4.0 and with the MPS2+ in AN521 mode.

karl-zh commented on samples/tfm_integration/README.rst:
```diff
@@ -0,0 +1,124 @@
+
+   mkdir build
+   cd build
+   cmake -G"Unix Makefiles" -DPROJ_CONFIG=`readlink -f ../ConfigCoreIPC.cmake` -DTARGET_PLATFORM=MUSCA_B1 -DBL2=false -DCOMPILER=GNUARM ../
```
Hi Kevin,

Thanks for your test on latest TFM master. Currently TFM project is doing memory optimization work which has removed some partitions from ConfigCoreIPC.cmake, but it is fine if we figure the ConfigRegressionIPC.cmake here. I will update the doc later.

microbuilder replied:
They are in the process of moving the config files to a dedicated folder as well, just as a FYI.

karl-zh replied:
Yes. Discussion ongoing. Anyway, per your suggestion, I have added the TFM commit we are using, that will be good for developer to reference.

microbuilder commented:
It looks like this is failing in CI due to the `install` folder not being present: https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/44029/5/tests

microbuilder commented:
> Needs to be rebased against master, and I will test again with QEMU 4.0 and with the MPS2+ in AN521 mode.

Confirmed working with QEMU 4.0 and the latest upstream trusted-firmware-m commit, with the caveat of changing the config file to [ConfigRegressionIPC.cmake](https://git.trustedfirmware.org/trusted-firmware-m.git/tree/ConfigRegressionIPC.cmake?id=bba856401c7eb5d66d170c3ba96b9cfa8e955a2a).

karl-zh commented:
> It looks like this is failing in CI due to the `install` folder not being present: https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/44029/5/tests

Maybe worth to upload a fake install folder with this sample, but should mention somewhere the veneer in this install folder have to be replaced.

Commit:
arm: mps_an521_nonsecure: Add TFM BL2 support

PSA level 1 requires secure boot. TF-M BL2 is the official
secure boot loader. It needs a BL2_HEADER_SIZE offset.

Align nonsecure address with TF-M's NS slot while TF-M BL2 enabled.

Signed-off-by: Karl Zhang <karl.zhang@linaro.org>

microbuilder reviewed (changes requested):
Given that other TF-M samples will be added shortly, the IPC sample app in this PR should be moved to `samples/tfm_integration/ipc_test`.

Other than that, and the need to rebase and address the documentation issues, it looks good.

carlescufi commented:
License-wise we should be fine since the files added to `ext/` use 3-clause BSD and that has already been approved by the license review board.

galak reviewed (changes requested):
s_veneers.o isn't going to be acceptable, we need it in source form, so a .S or .C.

karl-zh commented:
> s_veneers.o isn't going to be acceptable, we need it in source form, so a .S or .C.

@microbuilder This may require your script to compile TFM as external project. If you are going to create a new PR, is it possible to inclued the commits from this PR?

microbuilder commented:
> > s_veneers.o isn't going to be acceptable, we need it in source form, so a .S or .C.
> 
> @microbuilder This may require your script to compile TFM as external project. If you are going to create a new PR, is it possible to inclued the commits from this PR?

I'm looking at an alternative approach, and can update this PR with the build script I wrote for the PSA Level 1 app pre-merge, yes.

microbuilder commented:
This PR currently fails in CI due to the missing `Pycrypto` dependency, which is used by `imgtool.py` from TF-M to sign binary files. Since `Pycrypto` has been unmaintained for a number of years, a change request was added to TF-M migrating to the more modern and actively maintained `cryptography` module, which will also resolve the CI issues once merged.

See: https://review.trustedfirmware.org/c/trusted-firmware-m/+/1695

karl-zh commented:
Update TF-M and Zephyr to latest. TF-M commit: ffdcaa8e869ccc7b8f6631726f86cfa6323c4fee

Here is the script to sign TFM with Zephyr.bin together.
`#/bin/bash

TFM_ROOT=/home/user/trusted-firmware-m

cd ${TFM_ROOT}/build/app && /usr/bin/python3 ${TFM_ROOT}/bl2/ext/mcuboot/scripts/assemble.py --layout ${TFM_ROOT}/build/image_macros_preprocessed.c -s ${TFM_ROOT}/build/unit_test/tfm_s.bin -n ${TFM_ROOT}/zephyr_ns.bin -o ${TFM_ROOT}/build/tfm_full.bin
cd ${TFM_ROOT}/build/app && /usr/bin/python3 ${TFM_ROOT}/bl2/ext/mcuboot/scripts/imgtool.py sign --layout ${TFM_ROOT}/build/image_macros_preprocessed.c -k ${TFM_ROOT}/bl2/ext/mcuboot/root-rsa-3072.pem --align 1 -v 0.0.0+0 -H 0x400 ${TFM_ROOT}/build/tfm_full.bin ${TFM_ROOT}/build/tfm_sign.bin`

Config file for AN521 on MPS2+
`TITLE: Versatile Express Images Configuration File
 [IMAGES]
 TOTALIMAGES: 2 ;Number of Images (Max: 32)
 IMAGE0ADDRESS: 0x10000000
 IMAGE0FILE: \SW\mcuboot.bin ; BL2 bootloader
 IMAGE1ADDRESS: 0x10080000
 IMAGE1FILE: \SW\tfm_sign.bin ; TF-M example application binary blob`

microbuilder committed:
sample: tfm: Build TF-M with Zephyr

Add script to clone and build TF-M repo automaticly. This is a
preparation to concatenate TF-M output and Zephyr to single image for
sign.

Signed-off-by: Kevin Townsend <kevin@ktownsend.com>
Signed-off-by: Karl Zhang <karl.zhang@linaro.org>

Commit:
sample: tfm_integration: Involve TF-M repo

The files in ext/tfm folder are from TF-M repo, they will be compiled
before Zephyr as TF-M's output files are needed to combine with Zephyr.

As secure boot is required by PSA level 1, BL2 should enabled. TF-M
secure image and Zephyr non-secure image should be concatenated and
signed. The commands need to be align with the TF-M BL2 to create and
sign image: ${TFM_ROOT}/bl2/ext/mcuboot/MCUBoot.cmake

Signed-off-by: Karl Zhang <karl.zhang@linaro.org>

karl-zh commented:
> Update TF-M and Zephyr to latest. TF-M commit: ffdcaa8e869ccc7b8f6631726f86cfa6323c4fee
> 
> **Here is the script to sign TFM with Zephyr.bin together:**
> #/bin/bash
> 
> TFM_ROOT=/home/user/trusted-firmware-m
> 
> cd ${TFM_ROOT}/build/app && /usr/bin/python3 ${TFM_ROOT}/bl2/ext/mcuboot/scripts/assemble.py --layout ${TFM_ROOT}/build/image_macros_preprocessed.c -s ${TFM_ROOT}/build/unit_test/tfm_s.bin -n ${TFM_ROOT}/zephyr_ns.bin -o ${TFM_ROOT}/build/tfm_full.bin
> cd ${TFM_ROOT}/build/app && /usr/bin/python3 ${TFM_ROOT}/bl2/ext/mcuboot/scripts/imgtool.py sign --layout ${TFM_ROOT}/build/image_macros_preprocessed.c -k ${TFM_ROOT}/bl2/ext/mcuboot/root-rsa-3072.pem --align 1 -v 0.0.0+0 -H 0x400 ${TFM_ROOT}/build/tfm_full.bin ${TFM_ROOT}/build/tfm_sign.bin
> 
> **Config file "images.txt" for AN521 on MPS2+**
> 
> TITLE: Versatile Express Images Configuration File
> [IMAGES]
> TOTALIMAGES: 2 ;Number of Images (Max: 32)
> IMAGE0ADDRESS: 0x10000000
> IMAGE0FILE: \SW\mcuboot.bin ; BL2 bootloader
> IMAGE1ADDRESS: 0x10080000
> IMAGE1FILE: \SW\tfm_sign.bin ; TF-M example application binary blob

For now it can be merge tfm_s.bin and zephyr.bin automatically by setting "CONFIG_BUILD_WITH_TFM=y".

microbuilder commented:
Replaced by #19985
